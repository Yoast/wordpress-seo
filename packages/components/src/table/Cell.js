import PropTypes from "prop-types";
import styled from "styled-components";

/* Yoast dependencies */
import { colors } from "@yoast/style-guide";

/**
 * Separatifies an element.
 *
 * @returns {String} CSS
 */
export function separatify() {
	return `
		&::after {
			position:relative;
			display: inline-block;
			border-right: 2px solid ${ colors.$color_grey };
			padding-right: 40px;
			height: 60px;
			content: "";
		}

		@media screen and ( max-width: 1355px ) {
			&::after {
				padding-right: 24px;
			}
		}

		@media screen and ( max-width: 800px ) {
			&::after {
				height: 48px;
				padding-right: 16px;
			}
		}
	`;
}

/**
 * Make an element text nowrap and add ellipsis.
 *
 * @returns {String} CSS
 */
export function ellipsify() {
	return `
		overflow: hidden;
		text-overflow: ellipsis;
		white-space: nowrap;
	`;
}

/*
 * The "Cells" are flex-items, their "flex" property is the initial `0 1 auto`
 * that translates to: cannot grow, can shrink, initial width based on the content.
 * A cell can be changed on a case by case basis wrapping the cells in a
 * styled component.
 */
const CellBase = styled.span`
	font-size: 14px;
	padding-left: 40px;

	&:first-child {
		padding-left: 0;
	}

	&::before {
 		position: absolute;
 		left: -9999em;
 		top: -30px;
 		font-size: 1.286em;
 		line-height: 0;
 		${ props => props.headerLabel ? `content: "${ props.headerLabel }";` : "content: none;" }
 	}

 	${ props => props.separator ? separatify() : "" }
	${ props => props.ellipsis ? ellipsify() : "" }

	@media screen and ( max-width: 1355px ) {
		padding-left: 20px;
		${ props => props.hideOnTablet ? "display: none;" : "" }
	}

	@media screen and ( max-width: 800px ) {
		${ props => props.hideOnMobile ? "display: none;" : "" }
	}
`;

CellBase.propTypes = {
	children: PropTypes.any,
	hideOnMobile: PropTypes.bool,
	hideOnTablet: PropTypes.bool,
	separator: PropTypes.bool,
	headerLabel: PropTypes.string,
	ellipsis: PropTypes.bool,
};

CellBase.defaultProps = {
	hideOnMobile: false,
	hideOnTable: false,
	separator: false,
	ellipsis: false,
};

/*
 * Primary cell, the largest one in a row: can grow, cannot shrink, and the
 * initial width is 200 pixels. In the responsive view, can shrink.
 */
export const CellPrimary = styled( CellBase )`
	flex: 1 0 200px;

	@media screen and ( max-width: 800px ) {
		flex-shrink: 1;
	}
`;

/*
 * Cell with fixed width: cannot grow, cannot shrink, and the width based on
 * its content.
 */
export const CellFixedWidth = styled( CellBase )`
	flex: 0 0 auto;
`;

/*
 * Cell with a minimum width: can grow, cannot shrink, and the initial width
 * is 100 pixels.
 */
export const CellMinWidth = styled( CellBase )`
	flex: 1 0 100px;
`;

/*
 * Cell with icon: cannot grow, cannot shrink, and the width is fixed based on
 * its content. The height is smaller in the responsive view because icons are
 * smaller.
 */
export const CellIcon = styled( CellFixedWidth )`
	height: 60px;

	@media screen and ( max-width: 800px ) {
		height: 48px;
	}
`;

/**
 * Makes cells use headers as bold labels with a colon in the mobile responsive view.
 *
 * @param {ReactElement} cell The original cell.
 *
 * @returns {ReactElement} The cell with transformed headers.
 */
export function responsiveHeaders( cell ) {
	return styled( cell )`
		@media screen and ( max-width: 800px ) {
			&::before {
				float: left;
				min-width: 60px;
				line-height: inherit;
				font-weight: 700;
				${ props => props.headerLabel ? `content: "${ props.headerLabel }:";` : "content: none;" }
			}

			> span {
				line-height: inherit;
			}
		}
	`;
}
