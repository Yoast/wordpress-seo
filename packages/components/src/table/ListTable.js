import PropTypes from "prop-types";
import React from "react";
import styled from "styled-components";

/* Yoast dependencies */
import { colors } from "@yoast/style-guide";

const List = styled.ul`
	margin: 0;
 	padding: 0;
 	list-style: none;
 	position: relative;
 	width: 100%;

 	li:first-child {
		& > span::before {
			left: auto;
		}
	}
`;

List.propTypes = {
	children: PropTypes.any,
};

/**
 * Makes an element full-width in the mobile responsive view.
 *
 * @param {ReactElement} component The original element.
 * @returns {ReactElement} The element with full width responsive style.
 */
export function makeFullWidth( component ) {
	return styled( component )`
		@media screen and ( max-width: 800px ) {
			min-width: 100%;
			margin-top: 1em;
			padding-right: 0;
			padding-left: 0;
		}
	`;
}

/**
 * Creates a list table component.
 */
class ListTable extends React.Component {
	/**
	 * Constructor for the component.
	 *
	 * @param {Object} props The props to assign to the current component.
	 *
	 * @constructor
	 */
	constructor( props ) {
		super( props );
	}

	/**
	 * Retrieves the child nodes of the ListTable and ensures that single results are wrapped in an array.
	 *
	 * @returns {Array} The children to use within the ListTable.
	 */
	getChildren() {
		if ( this.props.children === 1 ) {
			return [ this.props.children ];
		}

		return this.props.children;
	}

	/**
	 * Renders the ListTable.
	 *
	 * @returns {ReactElement} The ListTable component.
	 */
	render() {
		const children = this.getChildren();

		return (
			<List role="list">{ children }</List>
		);
	}
}

/**
 * A zebrafied variant of the ListTable component.
 */
class ZebrafiedListTable extends ListTable {
	/**
	 * Constructor for the component. Also makes a shadow copy of the props that we can manipulate.
	 * @param {Object} props The props to assign to the current component.
	 *
	 * @constructor
	 */
	constructor( props ) {
		super( props );
		this.zebraProps = Object.assign( {}, props );
	}

	/**
	 * Zebrafies the child items of the ListTable.
	 *
	 * @returns {Array} Array containing the zebrafied rows.
	 */
	zebrafyChildren() {
		let children = this.props.children;

		// If there is only one child, there is no array, meaning there is no map method.
		if ( ! this.props.children.map ) {
			children = [ children ];
		}

		this.zebraProps.children = children.map( ( child, index ) => {
			return React.cloneElement( child, {
				background: ( index % 2 === 1 ) ? colors.$color_white : colors.$color_background_light,
				key: index,
			} );
		} );
	}

	/**
	 * Renders the zebrafied ListTable.
	 *
	 * @returns {ReactElement} The zebrafied ListTable.
	 */
	render() {
		this.zebrafyChildren();

		return ( <List role="list" { ...this.zebraProps } /> );
	}
}

ListTable.propTypes = {
	children: PropTypes.oneOfType( [
		PropTypes.arrayOf( PropTypes.node ),
		PropTypes.node,
	] ),
};

export { ListTable, ZebrafiedListTable };
