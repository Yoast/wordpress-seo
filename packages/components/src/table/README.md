# Usage guidelines for Table components
## Cells
- CellPrimary is meant to be the main cell in a row. There should be just one CellPrimary in each row.
- The responsiveHeaders function is meant to be used in combination with RowResponsiveWrap. This function ensures cells are full width in the responsive view.
- For the time being, headers should always be short and never break to a second line because they have `line-height: 0;`.
- CellIcon should only contain an icon, and therefore never use ellipsis.

