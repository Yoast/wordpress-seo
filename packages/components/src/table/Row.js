import PropTypes from "prop-types";
import styled from "styled-components";

/* Yoast dependencies */
import { colors } from "@yoast/style-guide";

// The Rows are flex containers.
export const Row = styled.li`
	background: ${ props => props.background };
	display: flex;
	min-height: ${ props => props.rowHeight };
	align-items: center;
	justify-content: space-between;
`;

Row.propTypes = {
	background: PropTypes.string,
	hasHeaderLabels: PropTypes.bool,
	rowHeight: PropTypes.string,
};

Row.defaultProps = {
	background: colors.$color_white,
	hasHeaderLabels: true,
};

/*
 * A row with responsive wrap allows children to wrap in new lines in the responsive view.
 * If the columns have headers, they're displayed as labels inside the column
 * content.
 */
export const RowResponsiveWrap = styled( Row )`
	@media screen and ( max-width: 800px ) {
		flex-wrap: wrap;
		align-items: flex-start;

		&:first-child {
			margin-top: ${ props => props.hasHeaderLabels ? "24px" : "0" };
		}

		// Use the column headers (if any) as labels.
		& > span::before {
			position: static;
			display: inline-block;
			padding-right: 0.5em;
			font-size: inherit;
		}
		& > span {
			padding-left: 0;
		}
	}
`;
