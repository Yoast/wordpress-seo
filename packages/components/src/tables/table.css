.yoast-table {
	font-size: var(--yoast-font-size-default);
	color: var(--yoast-color-default);
	line-height: 1.2;
	border-spacing: 0;
	border: var(--yoast-border-default);
	border-bottom: 0;
	width: 100%;
}

.yoast-table tbody tr:nth-child(odd) {
	background-color: #f9f9f9;
}

.yoast-table th {
	color: var(--yoast-color-dark);
	text-align: left;
	padding: 18px 12px;
	border-bottom: var(--yoast-border-default);
	white-space: nowrap;
	font-weight: var(--yoast-font-weight-bold);
}

.yoast-table td {
	padding: 18px 12px;
	border-bottom: var(--yoast-border-default);
}

.yoast-table th:first-child,
.yoast-table td:first-child {
	padding-left: 16px;
}

.yoast-table th:last-child,
.yoast-table td:last-child {
	padding-right: 16px;
}

td.yoast-table__image,
td.yoast-table__button {
	padding: 10px 18px 9px;
}

.yoast-table.yoast-table--nobreak td,
tr.yoast-table--nobreak td,
td.yoast-table--nobreak {
	white-space: nowrap;
}

th.yoast-table--primary {
	width: 100%;
}

td.yoast-table--nopadding {
	padding: 0 12px;
}
