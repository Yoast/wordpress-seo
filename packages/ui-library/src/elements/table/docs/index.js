export { default as component } from "./component.md";
export { default as tableBody } from "./table-body.md";
export { default as tableCell } from "./table-cell.md";
export { default as tableHead } from "./table-head.md";
export { default as tableHeader } from "./table-header.md";
export { default as tableRow } from "./table-row.md";
export { default as minimal } from "./table-variant-minimal.md";
