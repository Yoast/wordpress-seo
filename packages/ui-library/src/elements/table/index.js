import classNames from "classnames";
import PropTypes from "prop-types";
import { forwardRef } from "@wordpress/element";

const rowClassNameMap = {
	variant: {
		striped: "even:yst-bg-slate-50 odd:yst-bg-white",
		plain: "",
	},
};

/**
 * @param {JSX.node} children The content.
 * @param {string} [className] Optional class name.
 * @param {Object} [props] Optional table props.
 * @returns {JSX.Element} The element.
 */
const Cell = ( { children, className = "", ...props } ) => (
	<td className={ classNames( "yst-px-3 yst-py-4 yst-text-sm yst-text-slate-500", className ) } { ...props }>
		{ children }
	</td>
);

Cell.propTypes = {
	children: PropTypes.node.isRequired,
	variant: PropTypes.oneOf( Object.keys( rowClassNameMap.variant ) ),
	className: PropTypes.string,
};

/**
 * @param {JSX.node} children The content.
 * @param {string} [variant] Optional variant. See `rowClassNameMap.variant` for the options.
 * @param {string} [className] Optional class name.
 * @param {Object} [props] Optional table props.
 * @returns {JSX.Element} The element.
 */
const Row = ( { children, variant = "plain", className = "", ...props } ) => (
	<tr className={ classNames( rowClassNameMap.variant[ variant ], className ) } { ...props }>
		{ children }
	</tr>
);

Row.propTypes = {
	children: PropTypes.node.isRequired,
	variant: PropTypes.oneOf( Object.keys( rowClassNameMap.variant ) ),
	className: PropTypes.string,
};

/**
 * @param {JSX.node} children The content.
 * @param {string} [className] Optional class name.
 * @param {Object} [props] Optional table props.
 * @returns {JSX.Element} The element.
 */
const Header = ( { children, className = "", ...props } ) => (
	<th
		className={ classNames( "yst-px-3 yst-py-4 yst-text-left yst-text-sm yst-font-semibold yst-text-slate-900", className ) }
		{ ...props }
	>
		{ children }
	</th>
);

Header.propTypes = {
	children: PropTypes.node.isRequired,
	className: PropTypes.string,
};

/**
 * @param {JSX.node} children The content.
 * @param {string} [className] Optional class name.
 * @param {Object} [props] Optional table props.
 * @returns {JSX.Element} The element.
 */
const Head = ( { children, className = "", ...props } ) => (
	<thead className={ classNames( "yst-bg-slate-50", className ) } { ...props }>{ children }</thead>
);

Head.propTypes = {
	children: PropTypes.node.isRequired,
	className: PropTypes.string,
};

/**
 * @param {JSX.node} children The content.
 * @param {string} [className] Optional class name.
 * @param {Object} [props] Optional table props.
 * @returns {JSX.Element} The element.
 */
const Body = ( { children, className = "", ...props } ) => (
	<tbody className={ classNames( "yst-divide-y yst-divide-gray-200 yst-bg-white", className ) } { ...props }>{ children }</tbody>
);

Body.propTypes = {
	children: PropTypes.node.isRequired,
	className: PropTypes.string,
};

/**
 * @param {JSX.node} children The content.
 * @param {string} [className] Optional class name.
 * @param {Object} [props] Optional table props.
 * @returns {JSX.Element} The element.
 */
const Table = forwardRef( ( { children, className = "", ...props }, ref ) => (
	<div className="yst-overflow-hidden yst-shadow yst-ring-1 yst-ring-black yst-ring-opacity-5 yst-rounded-lg">
		<table className={ classNames( "yst-min-w-full yst-divide-y yst-divide-slate-300", className ) } { ...props } ref={ ref }>
			{ children }
		</table>
	</div>
) );

Table.propTypes = {
	children: PropTypes.node.isRequired,
	className: PropTypes.string,
};

Table.Head = Head;
Table.Head.displayName = "Table.Head";
Table.Body = Body;
Table.Body.displayName = "Table.Body";
Table.Header = Header;
Table.Header.displayName = "Table.Header";
Table.Row = Row;
Table.Row.displayName = "Table.Row";
Table.Cell = Cell;
Table.Cell.displayName = "Table.Cell";

export default Table;
