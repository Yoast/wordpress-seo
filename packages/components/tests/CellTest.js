import React from "react";
import renderer from "react-test-renderer";

import { CellPrimary, CellFixedWidth, CellMinWidth, CellIcon, responsiveHeaders } from "../src/table/Cell";

test( "the CellPrimary component matches the snapshot", () => {
	const component = renderer.create(
		<CellPrimary headerLabel="PrimaryCell">
			This is a primary cell
		</CellPrimary>
	);

	const tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );

test( "the CellFixedWidth component matches the snapshot", () => {
	const component = renderer.create(
		<CellFixedWidth headerLabel="FixedWidthCell">
			This is a cell with a fixed width
		</CellFixedWidth>
	);

	const tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );

test( "the CellMinWidth component matches the snapshot", () => {
	const component = renderer.create(
		<CellMinWidth headerLabel="MinWidthCell">
			This is a call with a min width
		</CellMinWidth>
	);

	const tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );

test( "the CellIcon component matches the snapshot", () => {
	const component = renderer.create(
		<CellIcon headerLabel="IconCell">
			Icon
		</CellIcon>
	);

	const tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );

test( "the CellPrimary with responsiveHeaders matches the snapshot", () => {
	const CellPrimaryResponsive = responsiveHeaders( CellPrimary );
	const component = renderer.create(
		<CellPrimaryResponsive headerLabel="responsiveHeaders">
			This is a primary cell with responsive headers
		</CellPrimaryResponsive>
	);

	const tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );
