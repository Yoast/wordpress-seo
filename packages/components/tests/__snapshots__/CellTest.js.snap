// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the CellFixedWidth component matches the snapshot 1`] = `
.c0 {
  font-size: 14px;
  padding-left: 40px;
}

.c0:first-child {
  padding-left: 0;
}

.c0::before {
  position: absolute;
  left: -9999em;
  top: -30px;
  font-size: 1.286em;
  line-height: 0;
  content: "FixedWidthCell";
}

.c1 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

@media screen and ( max-width:1355px ) {
  .c0 {
    padding-left: 20px;
  }
}

@media screen and ( max-width:800px ) {

}

<span
  className="c0 c1"
>
  This is a cell with a fixed width
</span>
`;

exports[`the CellIcon component matches the snapshot 1`] = `
.c0 {
  font-size: 14px;
  padding-left: 40px;
}

.c0:first-child {
  padding-left: 0;
}

.c0::before {
  position: absolute;
  left: -9999em;
  top: -30px;
  font-size: 1.286em;
  line-height: 0;
  content: "IconCell";
}

.c1 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.c2 {
  height: 60px;
}

@media screen and ( max-width:1355px ) {
  .c0 {
    padding-left: 20px;
  }
}

@media screen and ( max-width:800px ) {

}

@media screen and ( max-width:800px ) {
  .c2 {
    height: 48px;
  }
}

<span
  className="c0 c1 c2"
>
  Icon
</span>
`;

exports[`the CellMinWidth component matches the snapshot 1`] = `
.c0 {
  font-size: 14px;
  padding-left: 40px;
}

.c0:first-child {
  padding-left: 0;
}

.c0::before {
  position: absolute;
  left: -9999em;
  top: -30px;
  font-size: 1.286em;
  line-height: 0;
  content: "MinWidthCell";
}

.c1 {
  -webkit-flex: 1 0 100px;
  -ms-flex: 1 0 100px;
  flex: 1 0 100px;
}

@media screen and ( max-width:1355px ) {
  .c0 {
    padding-left: 20px;
  }
}

@media screen and ( max-width:800px ) {

}

<span
  className="c0 c1"
>
  This is a call with a min width
</span>
`;

exports[`the CellPrimary component matches the snapshot 1`] = `
.c0 {
  font-size: 14px;
  padding-left: 40px;
}

.c0:first-child {
  padding-left: 0;
}

.c0::before {
  position: absolute;
  left: -9999em;
  top: -30px;
  font-size: 1.286em;
  line-height: 0;
  content: "PrimaryCell";
}

.c1 {
  -webkit-flex: 1 0 200px;
  -ms-flex: 1 0 200px;
  flex: 1 0 200px;
}

@media screen and ( max-width:1355px ) {
  .c0 {
    padding-left: 20px;
  }
}

@media screen and ( max-width:800px ) {

}

@media screen and ( max-width:800px ) {
  .c1 {
    -webkit-flex-shrink: 1;
    -ms-flex-negative: 1;
    flex-shrink: 1;
  }
}

<span
  className="c0 c1"
>
  This is a primary cell
</span>
`;

exports[`the CellPrimary with responsiveHeaders matches the snapshot 1`] = `
.c0 {
  font-size: 14px;
  padding-left: 40px;
}

.c0:first-child {
  padding-left: 0;
}

.c0::before {
  position: absolute;
  left: -9999em;
  top: -30px;
  font-size: 1.286em;
  line-height: 0;
  content: "responsiveHeaders";
}

.c1 {
  -webkit-flex: 1 0 200px;
  -ms-flex: 1 0 200px;
  flex: 1 0 200px;
}

@media screen and ( max-width:1355px ) {
  .c0 {
    padding-left: 20px;
  }
}

@media screen and ( max-width:800px ) {

}

@media screen and ( max-width:800px ) {
  .c1 {
    -webkit-flex-shrink: 1;
    -ms-flex-negative: 1;
    flex-shrink: 1;
  }
}

@media screen and ( max-width:800px ) {
  .c2::before {
    float: left;
    min-width: 60px;
    line-height: inherit;
    font-weight: 700;
    content: "responsiveHeaders:";
  }

  .c2 > span {
    line-height: inherit;
  }
}

<span
  className="c0 c1 c2"
>
  This is a primary cell with responsive headers
</span>
`;
